### PR TITLE
Fixes relating to cross-device array copy

### PR DESCRIPTION
**Pull request by delta2323**
This PR implements
* multi_gpu decorator for cupy
* unittests for `cupy.copy`, `cupy.copyto` and `cupy.elementwise.copy` including multi-GPU ones.
* fixes for #438 
    * Contrary to what I commented there, I tried to support cross-device copy in non-UVA devices. But since I do not have such environment, I could not check the corresponding code.

### CODE REVIEW

**Comment by delta2323** [comment]
I was pointed out that current implementation already assumes UVA functions, so it is of no use that only `copy_from_device` supports non-UVA devices.

File: cupy/creation/from_data.py [hunk 1]
```diff
@@ -137,19 +136,10 @@ def copy(a):
...
-
-    # in-device copy
     newarray = cupy.empty_like(a)
-    if a.flags.c_contiguous:
-        newarray.data.copy_from(a.data, a.nbytes)
-    else:
-        elementwise.copy(a, newarray)
+    if not a.flags.c_contiguous:
+        a = ascontiguousarray(a)
+    newarray.data.copy_from(a.data, a.nbytes)
```
Comment (okuta) [inline cupy/creation/from_data.py]: Copy will be performed twice If same device and c_contiguous.

**Comment by okuta** [comment]
LGTM except comment.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**